### PR TITLE
Formalize spec

### PR DESCRIPTION
**Pull request by ivirshup**
Adding formal language and more details to the on disk specification for anndata.

The goal of this PR is to be more clear about what we expect to see from an AnnData formatted object on disk.

Eventually it would make sense to add some formatting to allow multiple versions of specifications, and to potentially separate prose and spec to separate documents.

## TODO

- [x] Remaining specifications
  - [x] AnnData shape restrictions
  - [x] String arrays
  - [x] Scalar types
  - [x] Nullable types
- [x] Check specifications against code
- [x] Get second pair of eyes for careful review

### CODE REVIEW

File: docs/fileformat-prose.md [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+>>> dict(f.attrs)
+{'encoding-type': 'anndata', 'encoding-version': '0.1.0'}
+```
+
+Using this information, we're able to dispatch onto readers for the different element types that you'd find in an anndata.
+
+### Element Specification
+
+* An element can be any object within the storage hierarchy (typically an array or group) with associated metadata
+* An element MUST have a string valued field `"encoding-type"` in it's metadata
```
Comment (giovp) [inline docs/fileformat-prose.md:54]:
```suggestion
* An element MUST have a string valued field `"encoding-type"` in its metadata
```

File: docs/fileformat-prose.md [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+* An element MUST have a string valued field `"encoding-version"` in it's metadata that can be evaluated to a version
```
Comment (giovp) [inline docs/fileformat-prose.md:55]:
```suggestion
* An element MUST have a string-valued field `"encoding-version"` in its metadata that can be evaluated to a version
```

File: docs/fileformat-prose.md (hunk 1 above)
Comment (giovp) [reply docs/fileformat-prose.md:54]: maybe `string-valued`?

File: docs/fileformat-prose.md [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+### AnnData specification (v0.1.0)
+
+* An `AnnData` object MUST be a group. 
+* The group's metadata MUST include entries: `"encoding-type": "anndata"`, `"encoding-version": "0.1.0"`.
+* An `AnnData` group MUST contain entries `"obs"` and `"var"`, which MUST be dataframes.
```
Comment (giovp) [inline docs/fileformat-prose.md:61]: yet they can be empty dataframes but with specified index? I would specify that

File: docs/fileformat-prose.md [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+as they have native equivalents in H5py {doc}`h5py:high/dataset` and Zarr {ref}`Arrays <zarr:tutorial_create>`.
+We can see an example of this with dimensionality reductions stored in the `obsm` group:
+
+```python
+>>> f["obsm"].visititems(print)
+X_pca <HDF5 dataset "X_pca": shape (38410, 50), type "<f4">
+X_umap <HDF5 dataset "X_umap": shape (38410, 2), type "<f4">
+
+>>> dict(f["obsm"]["X_pca"].attrs)
+{'encoding-type': 'array', 'encoding-version': '0.2.0'}
```
Comment (giovp) [inline docs/fileformat-prose.md:82]: here the encoding-version is 0.2.0 how is that related to the anndata version 0.1.0 ?

File: docs/fileformat-prose.md [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+where the kind and shape of the sparse array is defined in the `Group`'s attributes:
+
+```python
+>>> dict(f["X"].attrs)
+{'encoding-type': 'csr_matrix',
+ 'encoding-version': '0.1.0',
+ 'shape': array([38410, 27899])}
+```
+
+Inside the group are the three constituent arrays:
```
Comment (giovp) [inline docs/fileformat-prose.md:120]:
```suggestion
Inside the group there are three constituent arrays:
```

File: docs/fileformat-prose.md [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+<!-- ```bash
+$ h5ls 02_processed.h5ad/X
+data                     Dataset {41459314/Inf}
+indices                  Dataset {41459314/Inf}
+indptr                   Dataset {38411/Inf}
+``` -->
+
+### Sparse array specification (v0.1.0)
+
+* Each sparse array MUST be it's own group
```
Comment (giovp) [inline docs/fileformat-prose.md:138]:
```suggestion
* Each sparse array MUST be its own group
```

File: docs/fileformat-prose.md [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes',
+        'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes',
+        'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito',
+        'label_by_score'], dtype=object),
+ 'encoding-type': 'dataframe',
+ 'encoding-version': '0.2.0'}
+```
+
+These attributes identify the index of the dataframe, as well as the original order of the columns.
+Each column in this dataframe is encoded as it's own array.
```
Comment (giovp) [inline docs/fileformat-prose.md:164]:
```suggestion
Each column in this dataframe is encoded as its own array.
```

Comment (giovp) [reply docs/fileformat-prose.md:164]: also `DataFrame` or `dataframe` ?

File: docs/fileformat-prose.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+{'encoding-type': 'array', 'encoding-version': '0.2.0'}
+
+>>> dict(f["obs"]["cell_type"].attrs)
+{'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
+```
+
+### Dataframe Specification (v0.2.0)
+
+* A dataframe MUST be stored as a group
+* The group's metadata
```
Comment (giovp) [inline docs/fileformat-prose.md:177]:
```suggestion
* The group's metadata:
```

File: docs/fileformat-prose.md [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    * MUST contain the field `"_index"`, which is the name of the column to be used as an index
+    * MUST contain encoding metadata `"encoding-type: "dataframe"`, `"encoding-version": "0.2.0"`
+    * MUST contain `"column-order"` a string valued array denoting the order of column entries
```
Comment (giovp) [inline docs/fileformat-prose.md:180]: or `string-valued`

File: docs/fileformat-prose.md [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+* The group MUST contain a column for the index
+* Each entry in the group MUST correspond to an array with equivalent first dimensions
```
Comment (giovp) [inline docs/fileformat-prose.md:182]:
```suggestion
* Each entry in the group MUST correspond to an array with the equivalent first dimension
```

File: docs/fileformat-prose.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+* Each entry SHOULD share chunk sizes
+
+## Mappings
+
+Mappings are simply stored as `Group` s on disk.
```
Comment (giovp) [inline docs/fileformat-prose.md:187]:
```suggestion
Mappings are simply stored as `Group`'s on disk.
```

File: docs/fileformat-prose.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+[...]
+pca <HDF5 group "/uns/pca" (2 members)>
+pca/variance <HDF5 dataset "variance": shape (50,), type "<f4">
+pca/variance_ratio <HDF5 dataset "variance_ratio": shape (50,), type "<f4">
+[...]
+```
+
+### Mapping specifications (v0.1.0)
+
+* Each mapping MUST be it's own group
```
Comment (giovp) [inline docs/fileformat-prose.md:204]:
```suggestion
* Each mapping MUST be its own group
```

File: docs/fileformat-prose.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+These arrays encode the labels as small width integers (`codes`), which map to the original label set (`categories`).
+We store these two arrays separately.
+
+```python
+>>> categorical.visititems(print)
+categories <HDF5 dataset "categories": shape (22,), type "|O">
+codes <HDF5 dataset "codes": shape (38410,), type "|i1">
+```
+
+### Categorical array specifcation (v0.2.0)
```
Comment (will-moore) [inline docs/fileformat-prose.md:252]: specification

File: docs/fileformat-prose.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+>>> dict(f["uns/neighbors/params/metric"].attrs)
+{'encoding-type': 'string', 'encoding-version': '0.2.0'}
+```
+
+### Scalar specification (v0.2.0)
+
+* Scalars MUST be written as a 0 dimensional array
+* Numeric scalars
+    * MUST have `"encoding-type": "numeric-scalar"`, `"encoding-version": "0.2.0"` in their metadata
+    * MUST be a single numeric value, including boolean, unsigned integer, signed integer,  floating point, or complex floating point.
```
Comment (giovp) [inline docs/fileformat-prose.md:228]:
```suggestion
    * MUST be a single numeric value, including boolean, unsigned integer, signed integer,  floating point, or complex floating point
```

Comment (giovp) [reply docs/fileformat-prose.md:228]: there is never a dot at the end of the sentence (before or after) this is just to keep consistency

File: docs/fileformat-prose.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+codes <HDF5 dataset "codes": shape (38410,), type "|i1">
+```
+
+### Categorical array specifcation (v0.2.0)
+
+* Categorical arrays MUST be stored as a group
+* The group's metadata MUST contain the encoding metadata `"encoding-type": "categorical"`, `"encoding-version": "0.2.0"`
+* The group's metadata MUST contain the boolean valued field `"ordered"`, which indicates whether the categories are ordered
+* The group MUST contain an integer valued array named `"codes"`
+    * The `"codes"` array MAY contain signed integer values. If so, the code `-1` denotes a missing value.
```
Comment (giovp) [inline docs/fileformat-prose.md:258]:
```suggestion
    * The `"codes"` array MAY contain signed integer values. If so, the code `-1` denotes a missing value
```

File: docs/fileformat-prose.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+* The group MUST contain an array called `"categories"`
+
+## String arrays
+
+Arrays of strings are handled differently than numeric arrays since numpy doesn't really have a good way of representing arrays of unicode strings.
+`anndata` assumes strings are text like data, so are variable length.
```
Comment (giovp) [inline docs/fileformat-prose.md:264]:
```suggestion
`anndata` assumes strings are text like data, so they are variable length.
```

File: docs/fileformat-prose.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+>>> dict(categorical["categories"].attrs)
+{'encoding-type': 'string-array', 'encoding-version': '0.2.0'}
+```
+
+### String array specifications (v0.2.0)
+
+* String arrays MUST be stored in arrays
+* The arrays's metadata MUST contain the encoding metadata `"encoding-type": "string-array"`, `"encoding-version": "0.2.0"`
+* In `zarr`, string arrays MUST be stored using `numcodecs`' `VLenUTF8` codec.
+* In `HDF5`, string arrays MUST be stored using the variable length string data type, with a utf-8 encoding.
```
Comment (giovp) [inline docs/fileformat-prose.md:276]: here again dot at the end of sentence

File: docs/fileformat-prose.md [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+
+## Nullable integers and booleans
+
+We support IO with Pandas nullable integer and boolean arrays.
+We represent these on disk similar to `numpy` masked arrays, `julia` nullable arrays, or `arrow` validity bitmaps (see {issue}`504` for more discussion).
+That is, we store a indicator array (or mask) of null values alongside the array of all values.
```
Comment (giovp) [inline docs/fileformat-prose.md:282]:
```suggestion
That is, we store an indicator array (or mask) of null values alongside the array of all values.
```

File: docs/fileformat-prose.md [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+* The group MUST contain an integer valued array under the key `"values"`
+* The group MUST contain an boolean valued array under the key `"mask"`
+
+### Nullable boolean specifications (v0.1.0)
+
+* Nullable booleans MUST be stored as a group
+* The group's attributes MUST have contain the encoding metadata `"encoding-type": "nullable-boolean"`, `"encoding-version": "0.1.0"`
+* The group MUST contain an boolean valued array under the key `"values"`
+* The group MUST contain an boolean valued array under the key `"mask"`
+* The `"values"` and `"mask"` array MUST be the same shape
```
Comment (giovp) [inline docs/fileformat-prose.md:314]:
```suggestion
* The `"values"` and `"mask"` array MUST be of the same shape
```

**Review by giovp — CHANGES_REQUESTED** [review]
mostly typos

File: docs/fileformat-prose.md (hunk 13 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:252]: Thanks for catching that!

Oops, had left spell check toggled off. Thought it was weird I didn't get anything wrong ...

File: docs/fileformat-prose.md (hunk 4 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:82]: It's the version of the array encoding, not the anndata encoding. Those are independent.

File: docs/fileformat-prose.md (hunk 1 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:54]:
```suggestion
* An element MUST have a string-valued field `"encoding-type"` in its metadata
```

File: docs/fileformat-prose.md (hunk 3 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:61]:
```suggestion
* An `AnnData` group MUST contain entries `"obs"` and `"var"`, which MUST be dataframes (though this may only have an index with no columns).
```

File: docs/fileformat-prose.md (hunk 5 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:120]:
```suggestion
The group contains three arrays:
```

File: docs/fileformat-prose.md (hunk 11 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:187]:
```suggestion
Mappings are simply stored as `Group`s on disk.
```

File: docs/fileformat-prose.md (hunk 17 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:276]:
```suggestion
* In `zarr`, string arrays MUST be stored using `numcodecs`' `VLenUTF8` codec
* In `HDF5`, string arrays MUST be stored using the variable length string data type, with a utf-8 encoding
```

File: docs/fileformat-prose.md (hunk 19 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:314]:
```suggestion
* The `"values"` and `"mask"` arrays MUST be the same shape
```

File: docs/fileformat-prose.md (hunk 16 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:264]:
```suggestion
`anndata` assumes strings are text-like data, so uses a variable length encoding.
```

File: docs/fileformat-prose.md [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+
+>>> dict(f["obs"]["cell_type"].attrs)
+{'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
+```
+
+### Dataframe Specification (v0.2.0)
+
+* A dataframe MUST be stored as a group
+* The group's metadata:
+    * MUST contain the field `"_index"`, which is the name of the column to be used as an index
```
Comment (lazappi) [inline docs/fileformat-prose.md:178]: The `'_index'` above is `'Cell'` but that doesn't appear in `'column-order'`. I assume that's correct but I would state that's the case here and where the index appears in the order.

File: docs/fileformat-prose.md [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+    * MUST contain encoding metadata `"encoding-type: "dataframe"`, `"encoding-version": "0.2.0"`
+    * MUST contain `"column-order"` an array of strings denoting the order of column entries
+* The group MUST contain a column for the index
+* Each entry in the group MUST correspond to an array with equivalent first dimensions
+* Each entry SHOULD share chunk sizes
```
Comment (lazappi) [inline docs/fileformat-prose.md:183]: Where are chunk sizes stored, I can't see them above? (also maybe it's worth explaining what they are?)

File: docs/fileformat-prose.md [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+categories <HDF5 dataset "categories": shape (22,), type "|O">
+codes <HDF5 dataset "codes": shape (38410,), type "|i1">
+```
+
+### Categorical array specification (v0.2.0)
+
+* Categorical arrays MUST be stored as a group
+* The group's metadata MUST contain the encoding metadata `"encoding-type": "categorical"`, `"encoding-version": "0.2.0"`
+* The group's metadata MUST contain the boolean valued field `"ordered"`, which indicates whether the categories are ordered
+* The group MUST contain an integer valued array named `"codes"`
```
Comment (lazappi) [inline docs/fileformat-prose.md:257]: I know it's obvious to Python people but it might be worth stating these should start from 0

File: docs/fileformat-prose.md [hunk 23]
```diff
@@ -0,0 +1,318 @@
...
+```
+
+### String array specifications (v0.2.0)
+
+* String arrays MUST be stored in arrays
+* The arrays's metadata MUST contain the encoding metadata `"encoding-type": "string-array"`, `"encoding-version": "0.2.0"`
+* In `zarr`, string arrays MUST be stored using `numcodecs`' `VLenUTF8` codec
+* In `HDF5`, string arrays MUST be stored using the variable length string data type, with a utf-8 encoding
+
+## Nullable integers and booleans
```
Comment (lazappi) [inline docs/fileformat-prose.md:278]: Can you only do this with integers and booleans? What if you have a string DataFrame column with missing values?

Comment (ivirshup) [reply docs/fileformat-prose.md:278]: So, that's the plan, but hasn't been done yet due to the in-memory representation. At the time I created the issue it seemed like pandas was going to figure this out soon, but I don't believe that has happened.

https://github.com/scverse/anndata/issues/679

File: docs/fileformat-prose.md [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+
+### Dataframe Specification (v0.2.0)
+
+* A dataframe MUST be stored as a group
+* The group's metadata:
+    * MUST contain the field `"_index"`, which is the name of the column to be used as an index
+    * MUST contain encoding metadata `"encoding-type: "dataframe"`, `"encoding-version": "0.2.0"`
+    * MUST contain `"column-order"` an array of strings denoting the order of column entries
+* The group MUST contain a column for the index
+* Each entry in the group MUST correspond to an array with equivalent first dimensions
```
Comment (ivirshup) [inline docs/fileformat-prose.md:182]:
```suggestion
### Dataframe Specification (v0.2.0)

* A dataframe MUST be stored as a group
* The group's metadata:
    * MUST contain the field `"_index"`, whose value is the key of the array to be used as an index
    * MUST contain encoding metadata `"encoding-type": "dataframe"`, `"encoding-version": "0.2.0"`
    * MUST contain `"column-order"` an array of strings denoting the order of column entries
* The group MUST contain an array for the index
* The index array MUST NOT appear in 
* Each entry in the group MUST correspond to an array with equivalent first dimensions
```

File: docs/fileformat-prose.md (hunk 20 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:178]: Responded here: https://github.com/scverse/anndata/pull/882/files#r1086559577

File: docs/fileformat-prose.md (hunk 21 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:183]: The chunk sizes are a feature of the underlying zarr or hdf5 container – so the container handles this information. I'm a little reluctant to include explanations of the underlying containers here right now since it expands the scope.

File: docs/fileformat-prose.md [hunk 25]
```diff
@@ -0,0 +1,318 @@
...
+
+```python
+>>> categorical = f["obs"]["cell_type"]
+>>> dict(categorical.attrs)
+{'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
+```
+
+Discrete labels can be efficiently represented with categorical arrays (similar to `factors` in `R`).
+These arrays encode the labels as small width integers (`codes`), which map to the original label set (`categories`).
+We store these two arrays separately.
```
Comment (ivirshup) [inline docs/fileformat-prose.md:244]:
```suggestion
Discrete values can be efficiently represented with categorical arrays (similar to `factors` in `R`).
These arrays encode the values as small width integers (`codes`), which map to the original label set (`categories`).
Each entry in the `codes` array is the zero-based index of the encoded value in the `categories` array. 
To represent a missing value, a code of `-1` is used.
We store these two arrays separately.
```

File: docs/fileformat-prose.md (hunk 22 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:257]: I've addressed this a little here, but also above in the prose:https://github.com/scverse/anndata/pull/882/files#r1086580754

```suggestion
* The group MUST contain an integer valued array named `"codes"` whose maximum value is the number of categories - 1
```

File: docs/fileformat-prose.md (hunk 21 above)
Comment (lazappi) [reply docs/fileformat-prose.md:183]: Fair enough. If it's handled by the container maybe it doesn't need to be mentioned (or just add something to say that is the case)?

File: docs/fileformat-prose.md (hunk 24 above)
Comment (lazappi) [reply docs/fileformat-prose.md:182]: I think that's good apart from the missing end of the sentence.

Just for my understanding where is the index stored if it's not a column? Is it still in the array but named something else? Or should the sentence say something like "MUST contain the field `"_index"`, whose value is the key of the _group containing an array_ to be used as an index? It's probably obvious, just my lack of familiarly with HDF5.

Comment (ivirshup) [reply docs/fileformat-prose.md:182]: > I think that's good apart from the missing end of the sentence.

Oh, I think I meant to actually remove that bullet, since that is allowed (maybe it shouldn't be)

> Just for my understanding where is the index stored if it's not a column?

It's still stored in the group. For example:

```python
>>> for k, v in f["obs"].items(): print(k, ":", dict(v.attrs))
Cell : {'encoding-type': 'string-array', 'encoding-version': '0.2.0'}
cell_type : {'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
label_by_score : {'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
log1p_n_genes_by_counts : {'encoding-type': 'array', 'encoding-version': '0.2.0'}
...
```

It's just not included as a column of the dataframe since it's not included in `f["obs"].attrs["column-order"]`.

One could have an array be both an index and a value inside the dataframe. This is actually probably a nice organizational practice. However, this could be a problem if you instead have an index and a column with the same names, but different values.

File: docs/fileformat-prose.md (hunk 21 above)
Comment (ivirshup) [reply docs/fileformat-prose.md:183]:
```suggestion
* Each entry SHOULD share chunk sizes (in the HDF5 or zarr container)
```

I've clarified a bit. I think it should be easier to look up now.
